File: ui-automation/selenium-java/src/test/java/tests/ecommerce/BasicSearchTest.java
package tests.ecommerce;

import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;
import org.testng.Assert;
import pages.EcommercePage;
import utils.BaseTest;

/**
 * Test class for Search Functionality
 * Mirrors the basicSearch.lambda.cy.js Cypress test
 */
public class BasicSearchTest extends BaseTest {
    private EcommercePage ecommercePage;
    
    @BeforeMethod
    public void setUpTest() {
        ecommercePage = new EcommercePage();
        ecommercePage.navigateToHomepage();
    }
    
    @Test(description = "Should display results when searching for existing product")
    public void shouldDisplayResultsWhenSearchingForExistingProduct() {
        // Perform search for sony
        ecommercePage.performSearch("sony");
        
        // Verify URL contains search parameter
        Assert.assertTrue(ecommercePage.getCurrentUrl().contains("search=sony"), 
            "URL should contain search parameter");
        
        // Verify search results
        Assert.assertTrue(ecommercePage.isSonyVaioImageVisible(), 
            "Sony VAIO image should be visible");
        Assert.assertTrue(ecommercePage.areProductsVisible(), 
            "Product layout should exist");
    }
    
    @Test(description = "Should display no results message for invalid search")
    public void shouldDisplayNoResultsMessageForInvalidSearch() {
        // Perform search for invalid term
        ecommercePage.performSearch("invalid");
        
        // Verify URL and no results message
        Assert.assertTrue(ecommercePage.getCurrentUrl().contains("search=invalid"), 
            "URL should contain invalid search parameter");
        Assert.assertTrue(ecommercePage.isNoResultsMessageVisible(), 
            "No results message should be visible");
        Assert.assertTrue(ecommercePage.areProductThumbsNotPresent(), 
            "Product thumbs should not exist");
    }
    
    @Test(description = "Should handle special characters in search query")
    public void shouldHandleSpecialCharactersInSearchQuery() {
        // Perform search with special character
        ecommercePage.performSearch("\"");
        
        // Verify search results with special character
        Assert.assertTrue(ecommercePage.getCurrentUrl().contains("search=%22"), 
            "URL should contain encoded special character");
        Assert.assertTrue(ecommercePage.areProductsVisible(), 
            "Product layout should exist");
        
        // Verify result count is shown
        Assert.assertTrue(ecommercePage.isShowingResultsTextVisible(), 
            "Showing results text should be visible");
        String resultText = ecommercePage.getShowingResultsText();
        Assert.assertTrue(resultText.matches(".*Showing \\d+.*"), 
            "Result text should match pattern 'Showing [number]'");
    }
}

File: ui-automation/selenium-javascript/tests/ecommerce/basicSearch.test.js
const { expect } = require('chai');
const BaseTest = require('../../utils/BaseTest');
const EcommercePage = require('../../pages/EcommercePage');

describe('Search Functionality on eCommerce Playground', function() {
    let baseTest;
    let driver;
    let ecommercePage;

    beforeEach(async function() {
        baseTest = new BaseTest();
        driver = await baseTest.setUp();
        ecommercePage = new EcommercePage(driver);
        await ecommercePage.navigateToHomepage();
    });

    afterEach(async function() {
        await baseTest.tearDown();
    });

    it('should display results when searching for existing product', async function() {
        // Perform search for sony
        await ecommercePage.performSearch('sony');

        // Wait for search results to load
        await driver.sleep(2000);

        // Verify URL contains search parameter
        const currentUrl = await ecommercePage.getCurrentUrl();
        expect(currentUrl).to.include('search=sony');

        // Verify search results
        const isSonyVaioVisible = await ecommercePage.isSonyVaioImageVisible();
        expect(isSonyVaioVisible).to.be.true;

        const areProductsVisible = await ecommercePage.areProductsVisible();
        expect(areProductsVisible).to.be.true;
    });

    it('should display no results message for invalid search', async function() {
        // Perform search for invalid term
        await ecommercePage.performSearch('invalid');

        // Wait for search results to load
        await driver.sleep(2000);

        // Verify URL and no results message
        const currentUrl = await ecommercePage.getCurrentUrl();
        expect(currentUrl).to.include('search=invalid');

        const isNoResultsVisible = await ecommercePage.isNoResultsMessageVisible();
        expect(isNoResultsVisible).to.be.true;

        const areProductThumbsNotPresent = await ecommercePage.areProductThumbsNotPresent();
        expect(areProductThumbsNotPresent).to.be.true;
    });

    it('should handle special characters in search query', async function() {
        // Perform search with special character
        await ecommercePage.performSearch('"');

        // Wait for search results to load
        await driver.sleep(2000);

        // Verify search results with special character
        const currentUrl = await ecommercePage.getCurrentUrl();
        expect(currentUrl).to.include('search=%22');

        const areProductsVisible = await ecommercePage.areProductsVisible();
        expect(areProductsVisible).to.be.true;

        // Verify result count is shown
        const isShowingResultsVisible = await ecommercePage.isShowingResultsTextVisible();
        expect(isShowingResultsVisible).to.be.true;

        const resultText = await ecommercePage.getShowingResultsText();
        expect(resultText).to.match(/Showing \d+/);
    });
});
